### DISCUSSION TITLE
Local path dependencies during development

### DISCUSSION THREAD

**Discussion by finswimmer** [python-poetry/poetry#3646] [Show and tell]
**What's the task?**

I'm developing a tool `Tool` which relies on a library `Lib` which is under heavy development too. So I would like to add this library as a path dependency in editable mode during developing `Tool`. When publishing my `Tool`, `Lib` should be handled as a normal dependency.

A support by poetry for the above task is asked regularly, e.g. in https://github.com/python-poetry/poetry/issues/1168#issuecomment-717205031.

**Why does poetry not support it (yet)?**

poetry makes sure, that all defined dependencies are resolvable for the range of python versions defined in the `[tool.poetry.dependencies]` section. The result is stored in the `poetry.lock`, which makes it possible to create identical environments on different systems.

Defining different sources for a package makes it necessary to resolve dependencies for all sources, because even if the name is equal, the defined dependencies can differ. This can lead to conflicts in dependency resolution.

Even one can assume that dependencies are equal, there's still the problem how to tell poetry from which source it should install the package?

Ideas about how to solve this problem are very welcome :smile: 

**Is there a workaround?**
Yes, it is! (But maybe it's not suitable for all use cases.)

To get it work, one have to publish a minimal release of `Lib` to the place, where `Tool` can fetch it in production (pypi, private repository, git repository, ...). This minimal release must have defined all necessary dependencies.

Once this is done, add `Lib` as a dependency to `Tool` in the normal way. poetry can resolve dependencies and install `Lib`.

Now we can add the path to `Lib` to the environment variable [PYTHONPATH ](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH). Doing this, python will look at this place for module files when running a script.

How you set `PYTHONPATH` depends on your workflow. E.g. if you are using PyCharm you can [attach](https://www.jetbrains.com/help/pycharm/open-projects.html#multiple) the `Lib`  project to `Tool` project and PyCharm will do that for you.

If you are working within a terminal, it might be a good idea to create an `.env` file containing:

```
PYTHONPATH=/path/to/lib:$PYTHONPATH
```

and run `source .env` once when starting the terminal.

**_Note:_** The dependency resolution, getting package metadata and installing scripts given in the `[tool.poetry.scripts]` are made based on your published minimal release. So if these things changes you have publish a new release, so poetry can fetch up these changes.

**Update 2024-07-09:** In the meantime an alternative is to simply run `pip install -e /path/to/project` to temporary install the dependency in development mode.

**Comment by absassi** [comment]
> Defining different sources for a package makes it necessary to resolve dependencies for all sources, because even if the name is equal, the defined dependencies can differ. This can lead to conflicts in dependency resolution.

This only leads to a conflict if Poetry considers possible that both sources can be used _at the same time_, but that is not possible.
It is not too much different than a package having two versions (each version has its own set of dependencies, and may conflict with the other). The critical difference is that for versions, only one version is locked in the end, while to support this feature, Poetry will have to resolve and lock both cases of dev and non-dev environments.
The trivial solution is to just resolve everything twice if any dependency is specified differently in non-dev and dev sections. I believe the slowest part of resolution is gathering all the dependencies for each package, so with proper in-memory caching resolving twice shouldn't be a big performance penalty.
Obviously, there would be the need to either change the lock file format to support two separate resolved sets of dependencies (perhaps a new key for each package to say if the locked data refers to dev, non-dev or both), or simply create two lock files. Using a single file is much better, as normally it would be almost all the same for both, so things could be reused.

> Even one can assume that dependencies are equal, there's still the problem how to tell poetry from which source it should install the package?

I'd say `poetry install` selects the dev source, `poetry install --no-dev` selects the non-dev source.

**Reply by TBBle** [reply]
I think that dev/non-dev is just the right-now state of more general ['extras'](https://www.python.org/dev/peps/pep-0508/#extras) support, so it probably does not make sense to have one-file-per-X or boolean states, as those "2-state" situations are really just a specific case of "n-state" (where n is still likely to be small, but is combinational, so perhaps will go as high 16 distinct possible combinations of 4 extras).

Extras _tend_ to add dependencies, like 'dev' does now, but similarly to 'dev', they may flow through into _modifying_ dependencies, e.g., I create a Poetry package with a 'markdown' extra, and that extra requires the 'markdown' extra of some other package I used.

**Reply by finswimmer** [reply]
Hello @absassi,

> The critical difference is that for versions, only one version is locked in the end, while to support this feature, Poetry will have to resolve and lock both cases of dev and non-dev environments.

this is something I thought about too. Beside the performance issue you've mentioned, there is another problem: The locked version for the packages can be then different between dev and no-dev. Is this something users can live with? Does this irritate users? Or is this actually a no-go? I don't know ...

**Reply by absassi** [reply]
@finswimmer, they can be different, and the difference should be locked correctly.
Suppose `Tool` depends on `Lib-1.0.0` for non-dev, which in turn depends on `ThirdParty ^2.3.0` (thus some version like `2.3.4` will be locked for `Tool` non-dev). Now, the local version of `Lib` is being upgraded to support the shiny new API of `ThirdParty-3.0.0` and therefore now depends on `ThirdParty ^3.0.0`. I would expect that `Tool` locks `2.3.4` for non-dev (which is compatible with `Lib-1.0.0`) and locks `3.0.0` for dev (which is compatible with the currently in-development version of `Lib`).

**Reply by absassi** [reply]
In my answer, I missed an important point that @jonapich said in reply to the other answer:

> There's no point in locking the published version of libA because end users will obtain the constraint version and not the locked version. Developers don't want to use the constrained version, they want the one in the repository and poetry should basically ignore the version here.

Translating to example in the comment above, it means that Poetry would create only one lock file for `Tool` with `Lib` path and `ThirdParty-3.0.0`. There would be no lock file for non-dev with `Lib-1.0.0` and `ThirdParty-2.3.4`, because nobody needs that.

**Comment by abn** [comment]
An option here might to talk about a net-new feature rather than coerce existing capabilities. For example, we could maybe introduce a new command pair `attach`/`detach` (note this can be a plugin now too).

This might function as follows.

```sh
# add foobar as a dependency
poetry add foobar@1.0.0
````

```sh
# attach the local project into the active environment
poetry attach foobar /path/to/foobar
```

One option, if we want to keep things isolated, might be to create the following.
```console
$ tree .poetry/
.poetry/
├── lock
└── override.toml

0 directories, 2 file
$ cat .poetry/override.toml 
[tool.poetry.dependencies]
foobar = { path = "/path/to/foobar", develop = true }
```
The `.poetry/lock` file might contain the "local" lock file. The directory `.poetry` might be recommended ignored perhaps to avoid accidental commits.

```sh
# detach from active environment; remove temporary files
poetry detach foobar
```

All that said, I am not convinced that this really brings much to the table than a workflow where you do something like this;
```sh
poetry add foobar@1.0.0
poetry add -e /path/to/foobar
# do development what not
poetry add foobar@1.0.0
```

Is the ask in general that these "local" development dependencies be preserved in the `pyproject.toml` file, even in your vcs?

Alternatively `attach`/`detach`  could simply rely on something a local file like `.poetry/override.toml` to sort of alias the adds easier from the last example.

**Reply by jonapich** [reply]
I like the different file override approach, as it hides potentially repo-specific information from the pyproject.toml file

It seems like it would provide a definite solution to cases like #2693 where several python libraries are stashed in one repository. I would commit that file to the VCS.

If I may add... when using an overridden local dependency for development purposes, it should be used even if the version from its `pyproject.toml` doesn't match the constraint we will be publishing. This is to work better with automated CI/versioning systems where VCS tags and commit/merge comments are used to automatically bump the patch/minor/major versions.

**Reply by absassi** [reply]
> All that said, I am not convinced that this really brings much to the table than a workflow where you do something like this;
> ```
> poetry add foobar@1.0.0
> poetry add -e /path/to/foobar
> # do development what not
> poetry add foobar@1.0.0
> ```
> Is the ask in general that these "local" development dependencies be preserved in the pyproject.toml file, even in your vcs?

I would like to have that in the VCS, yes. So doing poetry add all the time isn't a solution. I prefer one command (`poetry install`) to setup my environment and the environment of my colleagues, not asking everyone to type 1+N commands where N is the number of libs I'm using and developing (which would then become one more script to add to VCS and maintain). Besides that, poetry add would modify pyproject.toml and poetry.lock and they would be committed by accident all the time. So this definitely brings something to the table.

I understand that some people may not want to add local paths to VCS, so not all devs need to have the sources of the libs, or could have them in other paths, and having this configuration in separate files gives this flexibility.

**Reply by TBBle** [reply]
I had a similar thought to @absassi , but I think maybe we need to distinguish "when working on a library, install it development mode" (which is what the quote from @abn is doing), and the "monorepo" workflow which @absassi appears to describe.

In the first workflow, only users working on that particular library need to use the pathed reference, and presumably, as demonstrated in @abn's snippet, they are doing so on a case-by-case basis. The path to the library is likely to be different for different users, and most users would probably not even have a checkout of the library. So pyproject.toml should _not_ contain the paths to the library (because they're user-specific, not project specific). Hence, the idea of isolating such library "overrides".

I feel that this is the workflow for which #1168 was originally opened, but am open to being corrected.

In the monorepo workflow, which a project I'm on is also using, so it's the reason I'm tracking these discussions the relative paths to the packages are static for the project, so they belong in the pyproject.toml. For this workflow, we have different issues:
* `poetry build` should not produce distribution packages with relative-path dependencies, but honour the version range in pyproject.toml for dependencies.
* `poetry install --no-dev` would like to install the required library *differently* to `poetry install`, i.e. either a non-editable install if still relative pathed, or a PyPI install by package version, not a pathed install.
  * I'd want the former, but I've seen a request on the Poetry issue tracker for the latter as well.
  * I assume this would be controlled by `poetry add`, but is impossible right now as this information is stored in poetry.lock, which doesn't distinguish dev/non-dev packages, but relies on the solver to determine one-true-install-setup for both.
* As a corollary to the above two, I'd want `poetry install` and `poetry lock` to fail if my pyproject.toml specifies a version match that is incompatible with the pathed dependency, so that once I bump the version in my library's pyproject.toml, I don't accidentally ship a version of my project that accepts an older, API-incompatible, version of my project.

There's also an outstanding bug in this latter workflow, where pathed projects do not upgrade correctly, as pathed dependencies don't check their versions. I had thought it was to do with the git-dependencies (which showed similar issues) which were fixed in 1.1.6, but 1.1.6 totally broke our pathed dependencies (they are failing to pull in some sub-dependencies, but not *all*, even though they are in poetry.lock correctly) so I haven't been able to test if this fix worked.

**Reply by brechtm** [reply]
> In the first workflow, only users working on that particular library need to use the pathed reference, and presumably, as demonstrated in @abn's snippet, they are doing so on a case-by-case basis. The path to the library is likely to be different for different users, and most users would probably not even have a checkout of the library. So pyproject.toml should _not_ contain the paths to the library (because they're user-specific, not project specific). Hence, the idea of isolating such library "overrides".

I would argue that for this use case it's overkill to involve Poetry. For the rare occasion where I needed to install a dependency in development mode, I just `pip install`ed it into the Poetry-managed virtual environment.

**Reply by TBBle** [reply]
That loses, for example, Poetry's view of what's in the venv, or what needs to be, if you're working with multiple venvs, multiple Python versions, or even have a problem and need to delete and recreate the venv. You can't just `poetry env delete X; poetry install` to get back to a pristine, known-good venv setup.

**Reply by brechtm** [reply]
You are correct, of course. However, at least for the times that I've needed to do that, that wasn't an issue because it was a very temporary setup to debug something. I'm sure there are situations where it would be useful to have Poetry support these, but they might be too rare to warrant that?

**Reply by TBBle** [reply]
The use-case as I understand it is not a short-term temporary setup to debug the other library, but when actively working on the other library, in order to test it in-place in its use-case. So it's local to the user, but not "just for 20 minutes". The original poster sounded like they would generally _leave_ it like that locally (as they tend to work on the two projects interweavedly), but don't want to publish the path-dependency by accident.

**Reply by jonapich** [reply]
Since `pip` doesn't have a `--dev` option when installing packages, it should be fine for the `dev-dependencies` section to refer to local repository folders, the same way you would list them in the `packages` sections of the `pyproject.toml`.

Also, there's no point in locking the published version of a module if it's local. For instance:

```
[deps]
libA = "*"

[dev-deps]
libA = { path = "../libA", develop = True }
```

There's no point in locking the published version of `libA` because end users will obtain the constraint version and not the locked version. Developers don't want to use the constrained version, they want the one in the repository and poetry should basically ignore the version here. In the off chance that you need a published version (for instance, to tackle some bug involving old versions), you can simply comment out the dev-deps line in your `pyproject.toml` (or pin it to an old published version) temporarily.

The above example _already works_  but it's not officially documented by Poetry. #2693 has a link to a public repository that uses the dep + dev-deps pattern successfully. I think Poetry should at least officialize this behavior, because it's very useful and doesn't seem too far off... and I don't want my whole repo to explode with some new poetry upgrade 😭

**Reply by TBBle** [reply]
That seems like a different (monorepo) use-case... where anyone who is developing on that package _also_ has `../libA` available by definition. (That's actually the use-case I have)

I'm also not clear how this works for non-developers, who would now be able to resolve _any_ `libA` version. Does this work?

```
[deps]
libA = ">3"

[dev-deps]
libA = { path = "../libA", develop = True }
```

I would expect, that if the above worked, I would get an error from `poetry install` if I accidentally had `../libA` checked out at an old revision, e.g., a mistake in the use of the `git submodules` command, or a user in a revision control system like Perforce or SVN where it's possible to update some part of the workspace and not another.

That said, I think the above would break `pip install mylib[dev]`, if Poetry was able to successfully export from that pyproject.toml, as I'm not sure where that `../libA` would point, but I'm fairly sure it wouldn't be right. (I don't know off-hand what Poetry currently does if you `poetry build` with pathed dependencies.)

**Reply by jonapich** [reply]
> That seems like a different (monorepo) use-case... where anyone who is developing on that package _also_ has `../libA` available by definition. (That's actually the use-case I have)

Yes, and in my case each library is published on its own even though they refer to the local version during development.

> I'm also not clear how this works for non-developers, who would now be able to resolve _any_ `libA` version. Does this work?

> 
> ```
> [deps]
> libA = ">3"
> 
> [dev-deps]
> libA = { path = "../libA", develop = True }
> ```

Yes, your example works as long as the local `libA`'s version matches the constraint (based on its `pyproject.toml` local file). Non-developers using `pip` will end up using `libA>3` from `pypi.org`.

If the local version doesn't match the constraint,  `poetry lock` will complain.

> That said, I think the above would break `pip install mylib[dev]`, if Poetry was able to successfully export from that pyproject.toml, as I'm not sure where that `../libA` would point

Indeed, it cannot work if one of the `extra` points to a local-only, dev dependency. But I could live with that limitation. To lift the limitation, we would need a new syntax for dependencies, something like `libA = { version = ">3", develop = "../libA" }`. For all user installs, use the versions. For dev installs (i.e. poetry install), use the one from the path.

> (I don't know off-hand what Poetry currently does if you `poetry build` with pathed dependencies.)

It works. `poetry build` doesn't include dependencies, it will package the sdist and wheel and include the `pyproject.toml` file. Then when you install it from `pip`, it will use the constraints from the dependencies section, and download `libA>3`.

**Reply by TBBle** [reply]
Excellent, thank you. We are currently using a monorepo with relative pathed dependencies, but hadn't yet started testing out how to make constraints and publishing work. As you mentioned, it's not really _documented_ how this would behave, and we haven't needed to publish yet (all our current use-cases are in the same monorepo for now), so it's good to know that we're not going to hit an unresolvable brick wall the moment we do start pushing thing to our local pypi index.

**Comment by MadJlzz** [comment]
Not sure if this can help the discussion here but on other langages I am used of having my dependencies loading first from the local `cache` of my package manager.

That being said, we could imagine that:
```shell
poetry publish --build --local-registry
```
builds and store the generated artifact in the local filesystem.

Running just behind
```shell
poetry lock && poetry install
```

would fetch this new package locally and install in in the `venv`

Working that way would also mean that the `pyproject.toml` dependency section would not change even if your are developing on a project that uses your lib changes locally. 

```pyproject.toml
[tool.poetry.dependencies]
# directory
my-package = "0.0.1-snapshot"
```
VS
```pyproject.toml
[tool.poetry.dependencies]
# directory
my-package = { path = "../my-package/", develop = true }
```

From the answer here: https://github.com/python-poetry/poetry/discussions/1135#discussioncomment-145756

**Reply by jonapich** [reply]
This is a good idea and solves a number of use cases, but it doesn't seem to play well with `develop=true` where you can edit the source of multiple local projects without extra steps.

In practice, all you need is a `my_package.pth` with the path to the local source of the package. This file is already created by poetry when `develop=true` is set, and placed in the `sites-package` folder.
